Sd 1433

cryogenian opened the pull request:
<img width="998" alt="2016-03-15 0 18 31" src="https://cloud.githubusercontent.com/assets/10245930/13760313/9c598d26-ea43-11e5-9d73-ef4e0b185831.png">
<img width="730" alt="2016-03-15 0 18 47" src="https://cloud.githubusercontent.com/assets/10245930/13760314/9c5a842e-ea43-11e5-9d52-53268d59c708.png">
<img width="609" alt="2016-03-15 0 18 56" src="https://cloud.githubusercontent.com/assets/10245930/13760315/9c5abe8a-ea43-11e5-98e3-5492d7fce9e8.png">
<img width="861" alt="2016-03-15 0 19 03" src="https://cloud.githubusercontent.com/assets/10245930/13760316/9c5b5a5c-ea43-11e5-9ef1-a8ef712f73e1.png">


We now that deck can have only one error card and one next action card. Why do we need it to be the same thing as other cards? It looks that next action card is much more similar with top level `+` than with `create child cell` submenu. Wouldn't it be better if `Notebook.Cell.Editor` knew about this uniqueness and handle error and next action card separately? 

Anyway this 

+ Fixes  https://slamdata.atlassian.net/browse/SD-1433. 
+ Formally fixes https://slamdata.atlassian.net/browse/SD-1435. Currently every cell can have its own next action card, I think this is incorrect behaviour for __3.0__. And because next cells are added to very bottom of notebook user can't see this after he pressed `Next` button, although, creating child of non-bottom cell is confusing too :smile: 
+ Fixes readonly mode for view mount output. 

Right now I'm working on making only one `Next` button active, removing top level `+` button.

@garyb review please

garyb commented:
Does there need to be a next button at all? I thought we just always had the next action card at the end. When we have the cell accordion, that will give a similar effect, since the next actions will only be visible when selected then.

jdegoes commented:
Indeed, there should be *no button*. A user can always flip through the deck to the next action card, no matter how deep the deck is. Choosing an option both morphs the next action card into the desired card, as well as adds a new next action card to the top of the deck.

Once we have the next action card, there is no need for the `+` button, either.

cryogenian commented:
Ok, do you think that it would be better if there wouldn't be `Next` type card? Just a bit altered current top level next card, right?

cryogenian commented:
I mean in code, in ui it must look like card

jdegoes commented:
> Ok, do you think that it would be better if there wouldn't be Next type card? Just a bit altered current top level next card, right?

Well, I feel that's an architectural decision. I have no opinion on it. When cards are sliding in a deck, it may be easier not to "fake it" and for it to be a real card type. But you know better than me. :smile: /cc @garyb

garyb commented:
I'm not sure either, I think maybe it would be easier to deal with as a "real" card in the long run though? Maybe we can fake it now while we still have the cell thing going on, but when the rest of the UI is card-ified it might not be as advantageous. Also by having it as a card means we don't have to do anything special to handle it in the accordion display logic.

cryogenian commented:
Ok, then I'll leave it to be cell with special handling in editor. Does it make sense?

garyb commented:
I think so, yeah. I see where the problem is now, I guess the actions it raises need to be special cased as we're not adding a child of the next action card, but of the "tail" of the current chain, etc. But that's how things are going to be as we proceed, so :+1:

cryogenian commented:
Thanks!

cryogenian commented:
@jdegoes should these next card/cell buttons be disabled or hidden when last cell has no result? For example for query cell that hasn't been played.

jdegoes commented:
@cryogenian Ideally, we would use views in query cards to avoid this issue. 

However, failing that, if a card is not producing anything, due to any reason (error or lack of user interaction), then the next card should be the error card which indicates what the problem is and suggests how to fix it.

cryogenian commented:
Does it mean that all cards in 3.0 are autoplayed? When user adds new jtable, it doesn't click __Play__ button to see results, right?

jdegoes commented:
> Does it mean that all cards in 3.0 are autoplayed? When user adds new jtable, it doesn't click Play button to see results, right?

Yes, we have to delete the notion of "play" from the cards. Instead, we can make the data flow as lazy as possible, evaluated only to the extent necessary. This means, for example, a query card would always produce a view (which would error if the view cannot be compiled); a cache card could be layered on a query card to compute the results and store it into a location (this would force the view); a visualization card layered onto a query card would also force the view, because the viz card can't do anything without the data.

cryogenian commented:
<img width="1280" alt="2016-03-21 17 50 57" src="https://cloud.githubusercontent.com/assets/10245930/13922490/9ab14c26-ef8d-11e5-815e-579a1a95daef.png">
<img width="1276" alt="2016-03-21 17 51 14" src="https://cloud.githubusercontent.com/assets/10245930/13922495/a0e1a492-ef8d-11e5-8dc6-025c6023908c.png">
<img width="1278" alt="2016-03-21 17 51 36" src="https://cloud.githubusercontent.com/assets/10245930/13922500/a522f240-ef8d-11e5-8cc0-7b12dcb9ec0f.png">

Switched output of `Download` and `Chart` cells from `Nothing` to `Just Blocked`. 
I've added these ^ `alert-info` for `Nothing` and `Just Blocked` cell outputs. Actually we won't need them after https://slamdata.atlassian.net/browse/SD-1436 and error card. 

I'm not sure how much this card should differ with others. Currently it shares all prisms machinery and have same type of slot (but with `top` as id) and very similar visual representation.

I'll update tests after #692 are merged. 

@garyb could you, please, review? or we should postpone it until tests are ready?

jdegoes commented:
In addition to an error card, we may need a **Working** card, which shows that work is being produced and the user is not able to move past the card until the output is complete. Eventually, we could use the per-card animation sequences to denote the work (but just generic gear animation for starters).

cryogenian commented:
I think this is ready for review
+ Rebased 
+ Tests are fixed (hopefully)

garyb commented:
I'm not sure if we need the "Please run next action" state - the fact that the next actions will appear when the pending cell is deleted or run is probably clue enough about that. Or if we are going to keep it I think we should probably make it more like a status message and reduce the text size to be in line with the other messages we show.

garyb commented on src/SlamData/Notebook/Editor/Component.purs:
```diff
@@ -295,9 +261,66 @@ peekCell cellId q = case q of
+      out <- map join $ H.query (CellSlot cellId) $ left (H.request GetOutput)
+      pure $ case out of
+        Nothing -> Just "Please, run the last card of deck"
+        Just Blocked -> Just "There is no available next actions"
```
"There _are_ no available next actions" :smile:

garyb commented on src/SlamData/Notebook/Editor/Component.purs:
```diff
@@ -295,9 +261,66 @@ peekCell cellId q = case q of
+    Just cellId -> do
+      out <- map join $ H.query (CellSlot cellId) $ left (H.request GetOutput)
+      pure $ case out of
+        Nothing -> Just "Please, run the last card of deck"
```
Assuming we keep the message, perhaps we should phrase it something like "Next actions will be made available once the last card has been run"? Without something like that, it might not be clear why we're asking the user to run the card.

garyb commented on src/SlamData/Notebook/Cell/Next/Component.purs:
```diff
@@ -0,0 +1,80 @@
+          , HE.onClick (HE.input_ (right ∘ AddCell cty))
+          ]
+          [ glyph (cellGlyph cty) ]
+      ]
```
Can we include a text label with these now? Text+Icons has great usability benefits over icons alone, and now we have giant buttons we should have the space to include that under each.

Maybe that wasn't in the wireframe, but I'd suggest we really should do it wherever possible (@jdegoes?)

cryogenian replied:
:+1: How these labels should look? Something like header?

garyb replied:
Actually, maybe something like this would be nice: 
![untitled-1](https://cloud.githubusercontent.com/assets/693642/13990697/26feba9e-f10d-11e5-9aa0-0d6790f3831b.png)

I had been thinking of having the text separate, but we've got space to inclue it inside the boxes even.

garyb commented on less/next-cell-card.less:
```diff
@@ -0,0 +1,22 @@
+            }
+        }
+    }
+}
```
Perhaps we should centrally align these icons too, they look a little strange to me aligned left as they are now.

`margin: auto` on the `ul` will probably do it, if we have fixed width `li`s

garyb commented:
Code all looks good to me :+1: hopefully some of the next action cell stuff will get simpler over time as we become more deck-like.

cryogenian commented:
Initial screen
<img width="1280" alt="2016-03-24 17 07 57" src="https://cloud.githubusercontent.com/assets/10245930/14019389/9ca26db4-f1e3-11e5-88ec-1579b76519f6.png">

Added markdown cell 
+ `border-bottom` breaks screen to _deck_ and next actions.
+ Switched to `h4` for messaging
<img width="1280" alt="2016-03-24 17 08 06" src="https://cloud.githubusercontent.com/assets/10245930/14019410/afc2f684-f1e3-11e5-89e5-fca07ec3efbb.png">

I don't like this. When there is only one choice for next action it looks very lonely. 
<img width="1280" alt="2016-03-24 17 08 18" src="https://cloud.githubusercontent.com/assets/10245930/14019454/edeb957e-f1e3-11e5-9364-5eab17ffbe93.png">

This is how blocked deck output is described. 
<img width="1280" alt="2016-03-24 17 08 45" src="https://cloud.githubusercontent.com/assets/10245930/14019476/0a984834-f1e4-11e5-836e-79740e5cae3f.png">

@garyb How does this look? Ok to merge?

garyb commented:
Looks great :+1: 

I see what you mean about the "lonely" next action though. I guess one option would be to always show all the buttons, but have a disabled state for the actions that aren't allowed (25% opacity or something)? I seem to recall at SlamFest we talked a bit about making it clear why actions were not allowed in some situations, so perhaps here we could use disabled state + `title` attribute explaining why the next action is not available here. I don't know though.

Thoughts @jdegoes / @beckyconning / @jonsterling?

I think this is good to go now though, we can revise any outstanding aesthetic/usability changes in additional PRs. Nice work :smile:

cryogenian commented:
Thanks for review!

beckyconning commented:
In the card ui spec we always show all actions, with the possible ones ordered to the top. If someone selects a non-possible action we explain why they can't do it.

jdegoes commented:
> In the card ui spec we always present all actions, with the possible ones ordered to the top. If someone selects a non-possible action we explain why they can't do it.

Exactly. Between that and level-of-detail adjustments and framing, I think the interface will improve a lot as we make more progress toward 3.0.
